neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/PersistentEObjectAdapter.java

```diff
@@ -43,113 +44,172 @@
+            .weakKeys()
+            .build();
 
-    /**
-     * This class should not be instantiated.
-     *
-     * @throws IllegalStateException every time
-     */
     private PersistentEObjectAdapter() {
-        throw new IllegalStateException("This class should not be instantiated");
+        throw Throwables.notInstantiableClass(getClass());
     }
 
     /**
-     * Returns the given {@code object} adapted in a specific {@code type}.
+     * Adapts the provided object as a {@code type} instance.
      *
-     * @param <T>             the type of the adapted object
-     * @param adaptableObject the object to adapt
-     * @param adapterType     the class in which the object must be adapted
+     * @param <T>    the type of the adapted object
+     * @param object the object to adapt
+     * @param type   the class in which the object must be adapted
      *
      * @return an adapted object in the given {@code type}, or {@code null} if the {@code object} cannot be assigned as
      * a {@code type}
      *
-     * @throws NullPointerException if the {@code type} is {@code null}
+     * @throws NullPointerException if any argument is {@code null}
      */
-    public static <T extends PersistentEObject> T getAdapter(@Nullable Object adaptableObject, Class<T> adapterType) {
-        if (isNull(adaptableObject)) {
-            return null;
-        }
-        checkNotNull(adapterType);
+    @Nonnull
+    private static <T extends PersistentEObject> T adapt(Object object, Class<T> type) {
+        checkNotNull(object, "object");
+        checkNotNull(type, "type");
 
         Object adapter = null;
-        if (adapterType.isInstance(adaptableObject)) {
-            adapter = adaptableObject;
+
+        if (type.isInstance(object)) {
+            adapter = object;
         }
-        else if (adaptableObject instanceof InternalEObject) {
-            adapter = ADAPTED_OBJECTS_CACHE.getIfPresent(adaptableObject);
-            if (isNull(adapter) || !adapterType.isAssignableFrom(adapter.getClass())) {
-                adapter = createAdapter(adaptableObject, adapterType);
-                ADAPTED_OBJECTS_CACHE.put((InternalEObject) adaptableObject, (PersistentEObject) adapter);
+        else if (InternalEObject.class.isInstance(object)) {
+            adapter = CACHE.get(object);
+            if (isNull(adapter) || !type.isAssignableFrom(adapter.getClass())) {
+                adapter = createAdapter(InternalEObject.class.cast(object), type);
+                CACHE.put(object, PersistentEObject.class.cast(adapter));
             }
         }
 
         if (isNull(adapter)) {
-            NeoLogger.warn("Unable to create a {0} adapter for this object of type {1}", adapterType.getSimpleName(), adaptableObject.getClass().getSimpleName());
+            throw new IllegalArgumentException(
+                    String.format("Unable to create a %s adapter for this object of type %s", type.getSimpleName(), object.getClass().getSimpleName()));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "object" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)